idx/metric_definition.go

```diff
@@ -0,0 +1,351 @@
+package idx
+
+import (
+	"bytes"
+	"crypto/md5"
+	"fmt"
+	"reflect"
+	"sort"
+	"strings"
+	"unsafe"
+
+	"github.com/raintank/schema"
+	log "github.com/sirupsen/logrus"
+)
+
+//go:generate msgp
+
+type mType uint8
+
+// MType represented by a uint8
+const (
+	MTypeUndefined mType = iota
+	MTypeGauge
+	MTypeRate
+	MTypeCount
+	MTypeCounter
+	MTypeTimestamp
+)
+
+// MetricName stores the name as a []uintptr to strings interned in an object store.
+// Each word is stored as a separate interned string without the '.'. The whole name
+// can be retrieved by calling MetricName.String()
+type MetricName struct {
+	nodes []uintptr
+}
+
+// Nodes returns the []uintptr of interned string addresses
+// for the MetricName
+func (mn *MetricName) Nodes() []uintptr {
+	return mn.nodes
+}
+
+// String returns the full MetricName as a string
+// using data interned in the object store
+func (mn *MetricName) String() string {
+	if len(mn.nodes) == 0 {
+		return ""
+	}
+
+	bld := strings.Builder{}
+	return mn.string(&bld)
+}
+
+func (mn *MetricName) string(bld *strings.Builder) string {
+	// get []int of the lengths of all of the mn.Nodes
+	lns, ok := IdxIntern.LenNoCprsn(mn.nodes)
+	if !ok {
+		// this should never happen, do what now?
+		return ""
+	}
+
+	// should be faster than calling IdxIntern.SetStringNoCprsn in a tight loop
+	var tmpSz string
+	szHeader := (*reflect.StringHeader)(unsafe.Pointer(&tmpSz))
+	first, _ := IdxIntern.ObjString(mn.nodes[0])
+	bld.WriteString(first)
+	for idx, nodePtr := range mn.nodes[1:] {
+		szHeader.Data = nodePtr
+		szHeader.Len = lns[idx+1]
+		bld.WriteString(".")
+		bld.WriteString(tmpSz)
+	}
+
+	return bld.String()
+}
+
+// setMetricName interns the MetricName in an
+// object store and stores the addresses of those strings
+// in MetricName.nodes
+func (mn *MetricName) setMetricName(name string) {
+	nodes := strings.Split(name, ".")
```

Comment:
I don't think we can assume properly formed metric names just yet because #1008 is not done yet
Until 1008 is fixed we can go ahead with this for further testing though, but not probably not for prod deployments, as users may send `...foo.bar` or `.foo..bar` etc metrics